packages/embark/src/lib/modules/console/index.ts

```diff
@@ -170,37 +175,41 @@ class Console {
     }
   }
 
-  private registerEmbarkJs() {
-    this.events.emit("runcode:register", "IpfsApi", IpfsApi, false);
-    this.events.emit("runcode:register", "Web3", Web3, false);
-    this.events.emit("runcode:register", "EmbarkJS", EmbarkJS, false);
-
-    EmbarkJS.Blockchain.done = true;
-    if (this.ipc.connected && !this.forceRegister) {
-      return;
-    }
-
-    this.events.once("code-generator-ready", () => {
-      const waitingForReady = setTimeout(() => {
-        this.logger.warn(__("Waiting for the blockchain connector to be ready..."));
-        // TODO add docs link to how to install one
-        this.logger.warn(__("If you did not install a blockchain connector, stop this process and install one"));
-      }, 5000);
-      this.events.request("blockchain:connector:ready", () => {
-        clearTimeout(waitingForReady);
-        this.events.request("code-generator:embarkjs:provider-code", (providerCode: string) => {
-          const func = () => {
-          };
-          this.events.request("runcode:eval", providerCode, func, true);
-          this.events.request("code-generator:embarkjs:init-provider-code", (initCode: string) => {
-            this.events.request("runcode:eval", initCode, () => {
-              this.events.emit("console:provider:done");
-              this.providerReady = true;
-            }, true);
-          });
+  private registerEmbarkJs(cb: Callback<null>) {
+    waterfall([
+      // wait for the VM to be setup
+      (next: any) => {
+        this.events.request("runcode:ready", next);
+      },
+      (next: any) => {
+        const waitingForReady = setTimeout(() => {
+          this.logger.warn(__("Waiting for the blockchain connector to be ready..."));
+          // TODO add docs link to how to install one
+          this.logger.warn(__("If you did not install a blockchain connector, stop this process and install one"));
+        }, 5000);
+        this.events.once("blockchain:connector:ready", () => {
+          clearTimeout(waitingForReady);
+          next();
         });
-      });
-    });
+      },
+      (next: any) => {
+        this.events.request("runcode:blockchain:connected", next);
+      },
+      // for every other case (including when asked for force), get the embarkjs
+      // provider code and eval it in the VM (either main running VM or console VM
+      // in the secondary process)
+      (next: any) => {
+        if (this.ipc.connected && !this.forceRegister) {
+          return next();
+        }
+        const connectCode = `EmbarkJS.Blockchain.connectConsole((err) => {
```

Comment:
It's not clear to that `Blockchain.connectConsole` returns anything, i.e. it seems to finish by invoking `cb` and `doneCb`, but perhaps I've misunderstood it.

Reply:
Ah nevermind, I see it's actually code that will be eval'd.